core/lib/types/src/l1_batch_commit_data_generator.rs

```diff
@@ -0,0 +1,74 @@
+use zksync_basic_types::{ethabi::Token, U256};
+
+use crate::{commitment::L1BatchWithMetadata, utils};
+
+pub trait L1BatchCommitDataGenerator
+where
+    Self: std::fmt::Debug + Send + Sync,
+{
+    fn l1_commit_data(&self, l1_batch_with_metadata: &L1BatchWithMetadata) -> Token;
+}
+
+#[derive(Debug, Clone)]
+pub struct RollupModeL1BatchCommitDataGenerator;
+
+#[derive(Debug, Clone)]
+pub struct ValidiumModeL1BatchCommitDataGenerator;
+
+impl L1BatchCommitDataGenerator for RollupModeL1BatchCommitDataGenerator {
+    fn l1_commit_data(&self, l1_batch_with_metadata: &L1BatchWithMetadata) -> Token {
+        Token::Tuple(rollup_mode_l1_commit_data(l1_batch_with_metadata))
+    }
+}
+
+impl L1BatchCommitDataGenerator for ValidiumModeL1BatchCommitDataGenerator {
+    fn l1_commit_data(&self, l1_batch_with_metadata: &L1BatchWithMetadata) -> Token {
+        Token::Tuple(validium_mode_l1_commit_data(l1_batch_with_metadata))
+    }
+}
+
+fn validium_mode_l1_commit_data(l1_batch_with_metadata: &L1BatchWithMetadata) -> Vec<Token> {
```

Comment:
You still need the `bytes pubdata` at the end of the encoding. It can have zero length, but you still need it. Otherwise, the ABI encoding wont match

Reply:
You're right! We filed an issue last week and will tackle it this week!

Reply:
Solved in this PR https://github.com/matter-labs/zksync-era/pull/1015/commits/d31d1a66e01a143136068b7f189f86963110d1a9

Reply:
Solved here https://github.com/lambdaclass/zksync-era/pull/136